index.js

```diff
@@ -1,135 +1,108 @@
-      var address = this.address().address
-      if (address === '::') address = '0.0.0.0'
-      console.log(`Listening on http://${address}:${this.address().port}`)
-    })
-}
+const Permissions = require('./permissions')
+const MultiMapeo = require('./multi-mapeo')
+const discoveryKey = require('./discoveryKey')
 
-function loadProjects (cb) {
-  mkdirp.sync('projects')
-  fs.readdir('projects', function (err, files) {
-    if (err) return cb(err)
-    var pending = 1
-    files.forEach(function (file) {
-      if (!/^[A-Fa-f0-9]{64}$/.test(file)) return
-      ++pending
-      var core = loadProject(file)
-      core.osm.ready(function () {
-        var ok = '[ OK ]'
-        var padding = new Array(process.stdout.columns||80 - file.length - ok.length).fill(' ').join('')
-        console.log(file + padding + ok)
-        if (!--pending) cb()
-      })
-    })
-    if (!--pending) cb()
-  })
+const PINO_OPTS = {
+  name: 'mapeo-web'
 }
 
-// Loads a project + starts swarming
-function loadProject (pid) {
-  if (projectCores[pid]) return projectCores[pid]
-
-  mkdirp.sync(path.join('projects', pid))
-  var dbdir = path.join('projects', pid, 'db')
-  var osm = Osm({ dir: dbdir, encryptionKey: pid })
-  var media = Blob(path.join('projects', pid, 'media'))
-
-  projectCores[pid] = new Mapeo(osm, media, { internetDiscovery: true })
-  projectCores[pid].sync.setName('mapeo-web') // TODO: some way for the operator to provide this
-  projectCores[pid].sync.listen()
-  projectCores[pid].sync.join(pid)
+module.exports = {
+  create
+}
 
-  // provides a reverse map of HASH => projectId
-  osm.core.use('filtermap', filtermap(pid))
+function create (opts) {
+  const mapeoWeb = new MapeoWeb(opts)
 
-  return projectCores[pid]
-}
+  mapeoWeb.initRoutes()
 
-function getProject (pid) {
-  return projectCores[pid]
+  return mapeoWeb
 }
 
-function removeProject (pid, cb) {
-  var core = getProject(pid)
-  if (!core) return process.nextTick(cb)
-  core.close(function () {
-    fs.rename(path.join('projects', pid), path.join('projects', 'dead-' + String(Math.random()).slice(2)), cb)
-    delete projectCores[pid]
-  })
-}
+class MapeoWeb {
+  constructor ({ logger = pino(PINO_OPTS), ...opts } = {}) {
+    const permissions = new Permissions({ ...opts, logger })
+    const multiMapeo = new MultiMapeo({ ...opts, logger })
+    const fastify = makeFastify({ logger })
+
+    this.logger = logger
+    this.permissions = permissions
+    this.multiMapeo = multiMapeo
+    this.fastify = fastify
+  }
+
+  initRoutes () {
+    this.fastify.register(require('fastify-websocket'))
+
+    // Leaving room for other major minor and patch versions.
+    this.fastify.get('/replicate/v1/:discoveryKey', { websocket: true }, async (connection, req, params) => {
+      const { discoveryKey } = params
+
+      this.permissions.getProjectKeyForDiscoveryKey(discoveryKey, (err, projectKey) => {
+        if (err || !projectKey) {
+          return connection.end('Invalid Key')
```

Comment:
Raw sockets don't have an error type. The most we can do is end the socket connection.

I don't think there's an easy way to detect this error in the protocol, it would lead to an error being emitted in parsing the protocol data on the client side.

@noffle Do you know if the mapeo protocol has a place to handle error messages in the handshake? We could simulate one of those if they exist.